### PR TITLE
fix: write file using existing newlines (fixes #115)

### LINKED ISSUE #115: On Windows, spurious ^M characters are added to Unix formatted files
Our team uses a `.gitattributes` file of
```
* -text  
```
This means files checked out on Windows will often contain only Unix line endings: just `^J`s (LF), and no `^M`s (CRs).

But when these files are reformatted using pydocstringformatter, the modified lines now end in `^M^J`, but only those lines.

Perhaps pydocstringformatter can sense the line ending characters of the lines it's modifying, and use them instead of introducing `^M`s where none were there previously?

I think this a better solution than adding a config option, but a config option would be nice too, to be explicit.

If there's interest, I can try to cobble a PR, that accomplishes this task.

### PR DESCRIPTION
**Pull request by rasa**
Here's my proposed solution. It should work in all cases, as it will default to the system default, if the file has different line endings (`file.newlines` returns a tuple).

I manually tested it with Unix, Dos, and Mac formatted files, and they all retained their line endings.

### CODE REVIEW

**Review by DanielNoord — CHANGES_REQUESTED** [review]
I don't think this fully solves the issue. If a repo contains files that are explicitly kept with `CRLF` and the tool is run on `Linux` in a `CI` it would still try to convert them based on the `os` newline.

Could you also add a test for this? I think you can copy one of the normal functional test but only need to make sure the newline is not `LF`. That should trigger a failure on the Windows or Linux tests.

**Comment by rasa** [comment]
Yes, I will add a test. Do you want me to add MacOS to the test matrix? A simple example is at
https://github.com/psf/black/blob/6140f04fe17ae7c542a29c207c9a45f6200f4e2c/.github/workflows/test.yml
which has evolved into
https://github.com/psf/black/blob/HEAD/.github/workflows/test.yml

**Comment by DanielNoord** [comment]
> Yes, I will add a test. Do you want me to add MacOS to the test matrix? A simple example is at
> https://github.com/psf/black/blob/6140f04fe17ae7c542a29c207c9a45f6200f4e2c/.github/workflows/test.yml
> which has evolved into
> https://github.com/psf/black/blob/HEAD/.github/workflows/test.yml

I think for this particular use case testing against Windows and Linux is enough, as they have different standard line endings and thus test the behaviour.

At some point I'll probably add macOS, but that's likely a task for another PR!

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 pydocstringformatter.egg-info
 dist/
 docs/_build
+**/__pycache__/
```
Comment (DanielNoord) [inline .gitignore:5]:
```suggestion
```

This should probably be ignored in your global `.gitignore` file.

**Comment by rasa** [comment]
So [this](https://github.com/DanielNoord/pydocstringformatter/blob/77b36a264afb6c03330cf52b5f4e36f8e0dd69a9/.pre-commit-config.yaml#L5) pre-commit hook is futzing my test files. I had committed these files in `tests/data/newlines` with the following newline formats:
```
dos.py       : dos
mac.py       : mac
mixed.py     : dos/unix
none.py      : none
unix.py      : unix
```
but the pre-commit hook rewrote them using these line endings:
```
dos.py       : dos
mac.py       : mac/unix
mixed.py     : dos/unix
none.py      : unix
unix.py      : unix
```
All the tests will still succeed, but breaks the tests, imo. Should I dynamically write these files out in the test's setup?

**Comment by DanielNoord** [comment]
> So [this](https://github.com/DanielNoord/pydocstringformatter/blob/77b36a264afb6c03330cf52b5f4e36f8e0dd69a9/.pre-commit-config.yaml#L5) pre-commit hook is futzing my test files. I had committed these files in `tests/data/newlines` with the following newline formats:
> 
> ```
> dos.py       : dos
> mac.py       : mac
> mixed.py     : dos/unix
> none.py      : none
> unix.py      : unix
> ```
> 
> but the pre-commit hook rewrote them using these line endings:
> 
> ```
> dos.py       : dos
> mac.py       : mac/unix
> mixed.py     : dos/unix
> none.py      : unix
> unix.py      : unix
> ```
> 
> All the tests will still succeed, but this breaks the tests, imo. Should I dynamically write these files out in the test's setup?

You can add the directory with the tests for this specific issue into the ignore of the pre-commit hook. 
https://github.com/DanielNoord/pydocstringformatter/blob/77b36a264afb6c03330cf52b5f4e36f8e0dd69a9/.pre-commit-config.yaml#L6
It's a regular expresion so adding `|dir/name/file/` should work!

File: tests/data/newlines/dos.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+var = """A multi-line
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply tests/data/newlines/dos.py:1]: Let me better explain myself.

By using the existing testing framework of input and `.out` files with a docstring that requires minor formatting (for example moving the quotes of a multi-line docstring to a new line we can fully check if we are inserting all new lines of the correct type.
Everything under https://github.com/DanielNoord/pydocstringformatter/tree/main/tests/data/format is ran automatically.
So `tests/data/format/newline/...` would be a good place to put these. 

Then copying and combining https://github.com/DanielNoord/pydocstringformatter/blob/main/tests/data/format/multi_line/module_docstring.py and https://github.com/DanielNoord/pydocstringformatter/blob/main/tests/data/format/ending_quote/module_docstring.py but changing the newlines would be a good fit as it is a very simple but clear change which would test all forms of new line insertions/removal.

[the remark this answers is not in the document]
Comment (rasa) [reply tests/data/newlines/dos.py:1]: Yikes! That sounds good, but perhaps you could implement? I'm a little lost, as I've only just started with this code base.

[the remark this answers is not in the document]
Comment (DanielNoord) [reply tests/data/newlines/dos.py:1]: Would it be okay for you if I set up the files like you currently did but then not do the line endings? I must confess I have no idea how to replicate those on my own Mac 😅 😄

**Review by rasa — COMMENTED** [review]
@DanielNoord Are you requesting this section be changed? All the tests pass, so I'm not sure what changes are needed.

**Comment by DanielNoord** [comment]
> @DanielNoord Are you requesting this section be changed? All the tests pass, so I'm not sure what changes are needed.

Sorry, this is probably turning out more difficult than you had imagined.

I have pushed a commit to your branch which includes correct tests for all 5 cases. You don't need to make any changes there. However, the `mixed` test case doesn't pass as the current solution is unable to determine the right newline. This means we probably need to rethink the solution.

**Comment by rasa** [comment]
> However, the `mixed` test case doesn't pass as the current solution is unable to determine the right newline. This means we probably need to rethink the solution.

Thanks for the help. Your solution is much simpler.

Re the mixed use case, files with non-standard newlines (`\n\r`, `\r\n\n`, etc.) are either text files that got corrupted in some way (such as older versions of Perforce) or they're not UTF-8 text files at all, but some other encoding, or even binary files. So let's just delete the use case.

Of course, a better solution would be to completely ignore processing these files. That way, non-UTF-8 files would never get corrupted. Thoughts?

**Comment by rasa** [comment]
Actually, I don't think this test framework is going to work for the "none" case, as the system will write the file using the system default newline character, so on Linux, `\n`, and on Windows, `\r\n`, and none.py.out has `\n`s hardcoded.

**Comment by rasa** [comment]
Ok, it's a little hacky, but all tests pass on Windows now. I think we can put this to bed. Agreed?

**Comment by DanielNoord** [comment]
> Ok, it's a little hacky, but all tests pass on Windows now. I think we can put this to bed. Agreed?

Hmm, yeah I'm not sure about the `mixed` and `none` examples. I think you might have been right before: they are so specific (and broken) that perhaps we shouldn't support them. Handling the different platforms by itself is already good enough for most users I think?

**Comment by rasa** [comment]
> Handling the different platforms by itself is already good enough for most users I think?

I'd leave the none/mixed tests in, even though they are edge cases, because you can never have too many tests (at least those that run instantaneously), and they do work. If we wanted, we could even add more extreme edge cases, such as utf8-bom, utf16, and non-text files with `\r`s and `\n`s in them.

**Comment by DanielNoord** [comment]
Thanks a lot @rasa!

I cleaned up the commit history and made some final changes. I'm going to make an immediate patch release with this included :)

File: pydocstringformatter/run.py [hunk 3]
```diff
@@ -85,7 +88,15 @@ def _format_file(self, filename: Path) -> bool:
                 filename_str = str(filename)
 
             if self.config.write:
-                with open(filename, "w", encoding="utf-8") as file:
+                if isinstance(newlines, tuple):
+                    newlines = newlines[0]
```
Comment (rasa) [inline pydocstringformatter/run.py:92]: @DanielNoord Your changes and comments look great. My only concern is if we process a utf-16, or non-text file, the file would most likely get corrupted. While I like the idea of automatically fixing text files that have non-standard newlines, we don't have any logic here to see if the file is really a text file, or not. Would checking for characters `\x00-\x08,\x0b-\x0c,\x0e-\x1f` tell us if the file is a binary file or not?

Perhaps I'm being overly cautious, but I think it's best to avoid corrupting files as much as possible, as the user may not even know the file is non-standard, or accidentally fed in the wrong files. Your thoughts?

Comment (DanielNoord) [reply pydocstringformatter/run.py:92]: Although that's a valid point I wonder how often such issues will occur in the wild. Either way, I would recommend to use non-write mode (so leave the `-w`) for any first time user or anybody who is not using a source control program.
If you're using `git` it's not too bad that one file is corrupted as you'll be able to easily revert anyway.

I think the benefit of being careful for very particular and unlikely examples in this case doesn't really weigh up against the benefits of keeping this a bit a simpler and doing some auto-formatting.

_Perhaps we should add the stuff about using `-w` with source control in the README._

Comment (rasa) [reply pydocstringformatter/run.py:92]: Thanks again for your help, and thoughtful changes. If it's ever decided to make it "binary aware", I confirmed that dos2unix [uses](https://sourceforge.net/p/dos2unix/dos2unix/ci/master/tree/dos2unix/common.c#l2034) the characters `\x00-\x08,\x0b,\x0e-\x1f` to determine if a file is binary.
